### PR TITLE
Issue57 - download PDF 

### LINKED ISSUE #57: Add tax receipt PDF generation into rabbitMQ
Add tax receipt PDF generation task into a task queue to avoid conflict with Heroku 30 sec request timeout problem.

Check out the code for task queue on CSV parser

### LINKED ISSUE #64: Tax receipt PDF generation bug with long list of items
It currently prints on a new entire page which it shouldn't. lol

Screenshot1
Screenshot2

### PR DESCRIPTION
**Pull request by omar2535**
Check if pdf is downloaded correctly, try for both single PDF and many PDF's. 
Should fix timeouts on heroku as well. 

closes #57 
closes #64

### CODE REVIEW

**Comment by omar2535** [comment]
There shouldn't  be 😞

File: app/views.py [hunk 1]
```diff
@@ -54,5 +61,64 @@ def poll_state(request):
...
+    #print(len(data))
+    task_id = 0
+    try:
+        task_id = request.build_absolute_uri().split("task_id=", 1)[1]                      #builds task id from URL
+    except:
+        return HttpResponseRedirect('/')
+    print(request.build_absolute_uri())                                                 #debugging purposes
+    print request.build_absolute_uri().split("task_id=", 1)[1]                          #debugging purposes
+    work = AsyncResult(task_id)                                                         #get the work from ID
+    if (work.ready()):                                                                  #check if it is complete
```
[the remark this answers is not in the document]
Comment (omar2535) [reply app/views.py]: @leesw98 
Is the celery worker running when you are being redirected to homepage?
Can't seem to re-create the redirect unless if celery isn't running.

**Comment by omar2535** [comment]
@leesw98 
Is the celery worker running when you are being redirected to homepage?
Can't seem to re-create the redirect unless if celery isn't running.

**Comment by leesw98** [comment]
Nevermind, it works now.

File: app/tasks.py [hunk 2]
```diff
@@ -12,6 +12,15 @@
...
 # Make sure worker is running, then tassk will be queued by worker to parse the data.
+from celery import Celery, current_task
+import csv, datetime, zipfile
+from .models import Donor, Donation, Item
+from celery import shared_task
+import datetime, StringIO, os
+from .utils import *
+from django.http import HttpResponseRedirect
+from django.http import HttpResponse
+from time import sleep
```
Comment (leesw98) [inline app/tasks.py]: Is this sleep used anywhere?

File: app/tasks.py [hunk 3]
```diff
@@ -12,6 +12,15 @@
 # Use 'celery -A reboot worker -l info'
 # Then in another terminal, run 'python manage.py runserver'
 # Make sure worker is running, then tassk will be queued by worker to parse the data.
+from celery import Celery, current_task
+import csv, datetime, zipfile
+from .models import Donor, Donation, Item
+from celery import shared_task
+import datetime, StringIO, os
```
Comment (leesw98) [inline app/tasks.py]: Same thing with StringIO?

File: app/templates/app/PollState.html [hunk 4]
```diff
@@ -75,39 +75,53 @@
...
+            window.location.href = "download/pdf" +"/?"+ "task_id=" + task_idg;
+        }
       }
     },500);
-   });
-  
+  })();
+
 </script>
+<IFRAME style="display:none" name="hidden-form"></IFRAME>
```
Comment (leesw98) [inline app/templates/app/PollState.html]: Please don't capitalize in html element names lol

File: app/views.py [hunk 5]
```diff
@@ -54,5 +61,52 @@ def poll_state(request):
...
+    if 'job' in request.GET:
+        job_id = request.GET['job']
+        job = AsyncResult(job_id)
+        data = job.result or job.state
+        context = {
+            'data': data,
+            'task_id': job_id,
+        }
+        return render(request, "app/PollState.html", context)
+    elif (request.method == 'POST'):
```
Comment (leesw98) [inline app/views.py]: () in Python condition is unnecessary

File: app/views.py [hunk 6]
```diff
@@ -54,5 +61,52 @@ def poll_state(request):
...
+    try:
+        task_id = request.build_absolute_uri().split("task_id=", 1)[1]                   #builds task id from URL
+    except:
+        return HttpResponseRedirect('/')
+
+    work = AsyncResult(task_id)                                                         #get the work from ID
+    if work.ready():                                                                    #check if task from worker is complete
+        try:
+            result = work.get(timeout=1)                                                #get result of work
+            try:
```
Comment (leesw98) [inline app/views.py]: Is there a reason there are two try statements? Having one try catch should catch all the problems most of the times.

File: app/tasks.py (hunk 3 above)
Comment (gauravnv) [reply app/tasks.py]: We should really clean up our requirements and imported packages

**Comment by omar2535** [comment]
xd. They were for testing purposes

File: app/templates/pdf/receipt.html [hunk 7]
```diff
@@ -79,14 +79,16 @@
         <title> Tax receipt for: {{ donor.customer_name }} </title>
         {% load static %}
 
+        <!-- DOESN"T WORK FOR CELERY
+        <link rel="stylesheet" type="text/css" href="app/templates/pdf/css/bootstrap.css">
```
Comment (leesw98) [inline app/templates/pdf/receipt.html]: then delete it lol

**Pull request merged**